content/blogposts/2023/slow-to-simd.md

```diff
@@ -0,0 +1,411 @@
+
+Exercise for the reader: why is it significant that we slice like `a[i:i+4:i+4]` rather than just `a[i:i+4]`?
+
+## Quantization
+
+We've improved execution speed of our code pretty dramatically at this point, but there is another dimension of performance that is relevant here: memory usage.
+
+(If you skipped the background section, this part might not make much sense.)
+
+In our situation, we're searching over vectors with 1536 dimensions. With 4-byte elements, this comes out to 6KiB per vector, and we generate roughly a million vectors per GiB of code. That adds up.
+
+When searching the vectors, they need to be held in RAM, which puts some serious memory pressure on our deployments. Moving the vectors out of memory would mean loading them from disk at search time, which is a no-go when performance is so critical.
+
+There are [plenty of ways](https://en.wikipedia.org/wiki/Dimensionality_reduction) to compress vectors, but we'll be talking about [_integer quantization_](https://huggingface.co/docs/optimum/concept_guides/quantization), which is relatively simple, but effective. The idea is to reduce the precision of the 4-byte `float32` vector elements by converting them to 1-byte `int8`s.
+
+I won't get into exactly _how_ we decide to do the translation between `float32` and `int8`, since that's a pretty [deep topic](https://huggingface.co/docs/optimum/concept_guides/quantization), but suffice it to say our function now looks like the following:
+
+```go
+func DotInt8BCE(a, b []int8) int32 {
+	if len(a) != len(b) {
+		panic("slices must have equal lengths")
+	}
+
+	sum := int32(0)
+	for i := 0; i < len(a); i += 4 {
+		aTmp := a[i : i+4 : i+4]
+		bTmp := b[i : i+4 : i+4]
+		s0 := int32(aTmp[0]) * int32(bTmp[0])
+		s1 := int32(aTmp[1]) * int32(bTmp[1])
+		s2 := int32(aTmp[2]) * int32(bTmp[2])
+		s3 := int32(aTmp[3]) * int32(bTmp[3])
+		sum += s0 + s1 + s2 + s3
+	}
+	return sum
+}
+```
+
+This change yields a 4x reduction in memory usage at the cost of some recall precision (which we carefully measured, but is irrelevant to this blog post).
+
+Re-running the benchmarks shows we suffer a perf hit from this change. Taking a look at the generated assembly (with `go tool compile -S`), there are a bunch of new instructions to convert `int8` to `int32`, so I expect that's the source of the slowdown. We'll make up for that in the next section.
+
+<Chart
+    rows={[
+        {name: "DotNaive", value: 941900},
+        {name: "DotUnroll4", value: 1286000},
+        {name: "DotBCE", value: 1407000},
+        {name: "DotInt8BCE", value: 1240000},
+    ]}
+/>
+
+
+## SIMD
+
+I always love an excuse to play with SIMD. And this problem seemed like the perfect nail for that hammer.
+
+For those unfamiliar, SIMD stands for "Single Instruction Multiple Data". Just like it's says, it lets you run an operation over a bunch of pieces of data with a single instruction. This is exactly what we want for calculating a dot product.
```

Comment:
Why is the SIMD function faster than the above? Can you motivate this a bit?

See "Web Aside OPT:SIMD" in CS:APP

e.g. If we look at agner fog (https://agner.org/optimize/instruction_tables.pdf) we know SIMD takes XXX cycles, while normal version using XXX cycles + parallelism is greater with SMID, etc.

Reply:
I added a little bit of motivation, but I don't think I have the room to get too deep here without burying the lede.